[Max Ang] iP

slackernoob opened the pull request:
## Duke Task Bot

Duke Task Bot frees your mind of having to remember the things you need to do. It's 

* text-based
* easy to learn 
* ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/slackernoob/ip).
2. double-click it.
3. add your tasks.
4. Let it manage your tasks for you. :star_struck:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public static void main(String[] args) throws DukeException {
        new Duke(System.getProperty("user.dir") + "/duke.txt").run();
    }
```

Commit:
Add Gradle support

slackernoob committed:
added Level-1: Greet, Echo, Exit

slackernoob committed:
Level-2: Add, List

slackernoob committed:
added Level-3: Mark as Done

slackernoob committed:
added Level-4: ToDo, Event, Deadline

slackernoob committed:
added A-TextUiTesting

slackernoob committed:
added Level-5: Handle Errors

slackernoob committed:
added Level-6: Delete

slackernoob committed:
Add save to hard disk feature

Task list is saved in a local variable, will be lost upon exit of program.

Unable to save task list for future use.

Saver class: help to save tasks in the hard disk automatically whenever task list changes.
data saved as a string into duke.txt file in the same directory.

slackernoob committed:
Add dates feature for tasks

Dates and timings were treated as a string previously.

This means that there is no formatting.

To enable a format for deadlines and dates, they are stored as LocalDate objects
within the respective classes Deadline and Event.

slackernoob committed:
Merge branch 'branch-Level-7'
Merge branch-Level-7 save feature into master branch

slackernoob committed:
Merge branch 'branch-Level-8'
Merge branch-Level-8 date feature into master branch

slackernoob committed:
Refactor code to include more classes: Storage, Ui, Parser, TaskList

Previously all logic clumped up in Duke class, confusing and messy.

Extracted out closely related code and split them into other classes.

slackernoob committed:
Organize classes into duke package

slackernoob committed:
Merge remote-tracking branch 'origin/add-gradle-support'
Merge add-gradle-support branch into master branch

slackernoob committed:
Add gradle support to build and run Duke

slackernoob committed:
Add JUnit tests for Commands class

slackernoob committed:
Add JavaDocs to public classes and methods

slackernoob committed:
Add minor edits to comply with coding standards

slackernoob committed:
Add find feature

slackernoob committed:
Merge branch 'branch-A-JavaDoc'
Merge branch-A-JavaDoc with JavaDocs into master branch

slackernoob committed:
Merge branch 'branch-A-CodingStandard'
Merge branch-A-CodingStandard to master branch

slackernoob committed:
Merge branch 'branch-Level-9'
Merge branch-Level-9 with find feature to master branch

cyiting commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Constructs the parser.
+     */
+    public Parser() {
```
I believe Java provides a default constructor, maybe there is no need for one here?

cyiting commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,183 @@
+ * and deal with interactions with the user.
+ */
+public class Ui {
+    private static final String spacesPrefix = "     ";
```
I like how you have static variables to define the spaces and the lines, but I think constants have to follow the screaming snake case?

cyiting commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,186 @@
+        String[] timings = timeRange.split(" to: ");
+
+        String start = timings[0];
+        String displayFormat = "MMM d yyyy";
+        String taskFormat = "yyyy-MM-dd";
```
Maybe these can be declared as static variables so that they can be used in your other methods as well?

cyiting reviewed (commented):
LGTM! I like how your chatbot is designed :) Your code is also easy to read and understand and is very clear. Javadoc comments are also well written, which helps with understanding the code.

slackernoob committed:
Add checkstyle gradle and plugin

slackernoob committed:
Merge branch 'A-CheckStyle'
Merge A-CheckStyle branch into master branch

slackernoob committed:
Start JavaFX Introduction

slackernoob committed:
Finish JavaFX Tutorial

slackernoob committed:
Add GUI to Duke

slackernoob committed:
Merge branch 'branch-Level-10'
Merge branch-Level-10 GUI feature to master branch

kennycjy commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,30 @@
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: "
+                + byDate.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
Good job on the wrapping :D

kennycjy commented on src/main/java/duke/DeadlineCommand.java:
```diff
@@ -0,0 +1,58 @@
+        ui.showNumTasks(taskList);
+    }
+
+    public String executeReturnString(TaskList taskList, Ui ui, Storage storage) throws DukeException {
```
Maybe do a javadoc to explain what this function does

kennycjy commented on src/main/java/duke/FindCommand.java:
```diff
@@ -0,0 +1,58 @@
+    public String executeReturnString(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        if (taskList.getSize() == 0) {
+            return ui.formatMessage("Task list is empty, can't find anything!");
+        } else {
```
Seemingly redundant else statement, can remove for better readability

kennycjy commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,192 @@
+        taskDescription = taskDescription.substring(0, taskDescription.length() - 1);
+        String timeRange = taskDetails[1].substring(6, taskDetails[1].length() - 1);
+        String[] timings = timeRange.split(" to: ");
+
```
Good use of new lines to improve readability

kennycjy reviewed (commented):
In general all looks pretty good and compliant

slackernoob committed:
Fix code style issues

slackernoob committed:
Add assertion

In Duke class the getResponse() method has a variable c that
should be an instance of Command.

Assertions are used to define assumptions about the program state
so that the runtime can verify them.

slackernoob committed:
Fix code quality issues

In classes Deadline and Event, date format is a magic string,
hence reimplemented them as String variables instead of magic
string.

slackernoob committed:
Add sort feature

Previously list command simply lists down the tasks in the order
that they were added to the list. This could be messy for the user.

As such, the sort feature allows the list to be sorted in
chronological order according to the dates. Todo tasks without
date would be listed at the very top of the list.

slackernoob committed:
Merge pull request #2 from slackernoob/branch-A-Assertions

Add assertion

slackernoob committed:
Merge pull request #3 from slackernoob/branch-A-CodeQuality

Fix code quality

slackernoob committed:
Merge pull request #4 from slackernoob/branch-BCD-Extension

Add sort feature

slackernoob committed:
Edit user guide

slackernoob committed:
Edit style and add javadocs

slackernoob committed:
Fix code style

slackernoob committed:
Modify application path
